packages/react-router/lib/router/define-route.ts

```diff
@@ -0,0 +1,510 @@
+  clientLoaderHydrate: true,
+  HydrateFallback() {
+    return null;
+  },
+  meta({ loaderData }) {
+    expectEqual<typeof loaderData, 2 | undefined>(true);
+    return [];
+  },
+  Component({ loaderData }) {
+    expectEqual<typeof loaderData, 2>(true);
+    return null;
+  },
+  ErrorBoundary({ loaderData }) {
+    expectEqual<typeof loaderData, 2 | undefined>(true);
+    return null;
+  },
+});
+
+// Loader data: client + hydrate + hydratefallback -> client
+defineRoute({
+  async clientLoader({ serverLoader }) {
+    expectEqual<typeof serverLoader, undefined>(true);
+    return 2 as const;
+  },
+  clientLoaderHydrate: true,
+  HydrateFallback() {
+    return null;
+  },
+  meta({ loaderData }) {
+    expectEqual<typeof loaderData, 2 | undefined>(true);
+    return [];
+  },
+  Component({ loaderData }) {
+    expectEqual<typeof loaderData, 2>(true);
+    return null;
+  },
+  ErrorBoundary({ loaderData }) {
+    expectEqual<typeof loaderData, 2 | undefined>(true);
+    return null;
+  },
+});
+
+// Loader data: client + hydrate + -> client
+defineRoute({
+  async clientLoader({ serverLoader }) {
+    expectEqual<typeof serverLoader, undefined>(true);
+    return 2 as const;
+  },
+  clientLoaderHydrate: true,
+  meta({ loaderData }) {
+    expectEqual<typeof loaderData, 2 | undefined>(true);
+    return [];
+  },
+  Component({ loaderData }) {
+    expectEqual<typeof loaderData, 2>(true);
+    return null;
+  },
+  ErrorBoundary({ loaderData }) {
+    expectEqual<typeof loaderData, 2 | undefined>(true);
+    return null;
+  },
+});
+
+// action: neither, server, client, both
+
+// Action data: no actions -> undefined
+defineRoute({
+  Component({ actionData }) {
+    expectEqual<typeof actionData, undefined>(true);
+    return null;
+  },
+  ErrorBoundary({ actionData }) {
+    expectEqual<typeof actionData, undefined>(true);
+    return null;
+  },
+});
+
+// Action data: server -> server
+defineRoute({
+  serverAction() {
+    return 1;
+  },
+  Component({ actionData }) {
+    expectEqual<typeof actionData, 1 | undefined>(true);
+    return null;
+  },
+  ErrorBoundary({ actionData }) {
+    expectEqual<typeof actionData, 1 | undefined>(true);
+    return null;
+  },
+});
+
+// Action data: client -> client
+defineRoute({
+  clientAction({ serverAction }) {
+    expectEqual<typeof serverAction, undefined>(true);
+    return 2;
+  },
+  Component({ actionData }) {
+    expectEqual<typeof actionData, 2 | undefined>(true);
+    return null;
+  },
+  ErrorBoundary({ actionData }) {
+    expectEqual<typeof actionData, 2 | undefined>(true);
+    return null;
+  },
+});
+
+// Action data: server + client -> client
+defineRoute({
+  serverAction() {
+    return 1;
+  },
+  clientAction() {
+    return 2;
+  },
+  Component({ actionData }) {
+    expectEqual<typeof actionData, 2 | undefined>(true);
+    return null;
+  },
+  ErrorBoundary({ actionData }) {
+    expectEqual<typeof actionData, 2 | undefined>(true);
+    return null;
+  },
+});
```

Comment:
@brophdawg11 and I were discussing this. Seems like `HydrateFallback` and `clientLoader.hydrate` would also affect types for `actionData`, same as they do for `loaderData`